File: backend/internal/service/security.py
from __future__ import annotations

import hashlib
from typing import Any
from uuid import UUID

from dao.models.security import SecurityScan
from dao.repositories.security import SecurityScanRepository
from pkg.apis.security.models import SecurityRequest, SecurityResult
from pkg.config.settings import Settings
from pkg.security.detectors import DETECTOR_REGISTRY
from pkg.security.evaluation import EvaluationReport, evaluate_scores
from pkg.security.pipeline import SecurityPipeline

from internal.telemetry.logging import get_logger
from internal.telemetry.metrics import SpiderMetrics

logger = get_logger("spider-security")


def prompt_hash(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


class SecurityService:
    def __init__(
        self,
        pipeline: SecurityPipeline,
        scans: SecurityScanRepository,
        settings: Settings,
        metrics: SpiderMetrics,
    ) -> None:
        self.pipeline = pipeline
        self.scans = scans
        self.settings = settings
        self.metrics = metrics

    async def inspect(
        self,
        text: str,
        *,
        request_id: UUID | None = None,
        source: str | None = None,
        user_id: UUID | None = None,
        model: str | None = None,
        metadata: dict[str, Any] | None = None,
        persist: bool = True,
        worker_id: str | None = None,
    ) -> tuple[SecurityResult, SecurityScan | None]:
        request = SecurityRequest(
            text=text,
            source=source,
            user_id=user_id,
            model=model,
            metadata=metadata or {},
        )
        if request_id is not None:
            request.request_id = request_id

        result = await self.pipeline.inspect(request)
        digest = prompt_hash(text)
        stored_text = text if self.settings.persist_prompt_content else None

        log_payload: dict[str, Any] = {
            "request_id": str(result.request_id),
            "decision": result.decision.value,
            "score": result.score,
            "detector": result.metadata.get("detector"),
            "threshold": result.metadata.get("threshold"),
            "latency_ms": result.total_latency_ms,
            "prompt_hash": digest,
            "prompt_length": len(text),
            "model": model,
        }
        if self.settings.log_prompt_content:
            log_payload["prompt"] = text
        logger.info("security_scan", **log_payload)

        detector_name = str(result.metadata.get("detector", "unknown"))
        detector_latency = (
            result.detector_results[0].latency_ms if result.detector_results else result.total_latency_ms
        )
        self.metrics.observe_scan(
            decision=result.decision.value,
            pipeline_latency_ms=result.total_latency_ms,
            chunks=result.chunks_scanned,
            detector_name=detector_name,
            detector_score=result.score,
            detector_latency_ms=detector_latency,
        )

        scan: SecurityScan | None = None
        if persist:
            scan = await self.scans.create_from_result(
                result,
                prompt_hash=digest,
                prompt_length=len(text),
                prompt_text=stored_text,
                user_id=user_id,
                worker_id=worker_id,
            )
            logger.info("security_scan_persisted", scan_id=str(scan.id), request_id=str(result.request_id))
        return result, scan

    def list_detectors(self) -> list[dict[str, str]]:
        implemented = [
            {"name": name, "status": "implemented", "warning": "rule-based is development/testing only"}
            if name == "rule-based"
            else {"name": name, "status": "implemented", "warning": ""}
            for name in DETECTOR_REGISTRY
        ]
        prepared = [
            {"name": "prompt-shield", "status": "prepared", "warning": "not implemented"},
            {"name": "flan-t5", "status": "prepared", "warning": "not implemented"},
            {"name": "transformer", "status": "prepared", "warning": "not implemented"},
            {"name": "remote", "status": "prepared", "warning": "not implemented"},
            {"name": "ensemble", "status": "prepared", "warning": "not implemented"},
        ]
        return implemented + prepared

    async def evaluate(
        self,
        samples: list[dict[str, Any]],
        *,
        threshold: float | None = None,
        target_fprs: list[float] | None = None,
    ) -> EvaluationReport:
        scores: list[float] = []
        labels: list[bool] = []
        for sample in samples:
            result, _scan = await self.inspect(
                sample["text"],
                source="evaluate",
                persist=False,
            )
            scores.append(result.score)
            labels.append(bool(sample["is_injection"]))
        used_threshold = (
            threshold if threshold is not None else float(self.pipeline.policy.threshold or 0.5)
        )
        return evaluate_scores(
            labels,
            scores,
            threshold=used_threshold,
            target_fprs=target_fprs or (0.0005, 0.001, 0.005, 0.01),
        )

File: backend/internal/service/security.go
package service

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"log/slog"

	"github.com/google/uuid"
	"github.com/spider/spider/dao/store"
	"github.com/spider/spider/internal/telemetry"
	"github.com/spider/spider/pkg/apis"
	"github.com/spider/spider/pkg/config"
	"github.com/spider/spider/pkg/security/detectors"
	"github.com/spider/spider/pkg/security/evaluation"
	"github.com/spider/spider/pkg/security/pipeline"
)

func PromptHash(text string) string {
	sum := sha256.Sum256([]byte(text))
	return hex.EncodeToString(sum[:])
}

type SecurityService struct {
	Pipeline *pipeline.Pipeline
	Policy   *PolicyService
	Scans    SecurityScanStore
	Settings *config.Settings
	Metrics  *telemetry.Metrics
}

func (s *SecurityService) Inspect(ctx context.Context, text string, opts InspectOptions) (apis.SecurityResult, *store.SecurityScan, error) {
	if s.Policy != nil {
		_ = s.Policy.ApplyActivePolicy(ctx)
	}
	request := apis.SecurityRequest{
		RequestID: opts.RequestID,
		Text:      text,
		Source:    opts.Source,
		Model:     opts.Model,
		Metadata:  opts.Metadata,
	}
	if request.RequestID == "" {
		request.RequestID = uuid.New().String()
	}
	if opts.UserID != nil {
		uid := opts.UserID.String()
		request.UserID = &uid
	}

	result := s.Pipeline.Inspect(ctx, request)
	digest := PromptHash(text)

	logAttrs := []any{
		"request_id", result.RequestID,
		"decision", result.Decision,
		"score", result.Score,
		"detector", result.Metadata["detector"],
		"threshold", result.Metadata["threshold"],
		"latency_ms", result.TotalLatencyMs,
		"prompt_hash", digest,
		"prompt_length", len(text),
	}
	if opts.Model != nil {
		logAttrs = append(logAttrs, "model", *opts.Model)
	}
	if s.Settings.LogPromptContent {
		logAttrs = append(logAttrs, "prompt", text)
	}
	slog.Info("security_scan", logAttrs...)

	detectorName := "unknown"
	if v, ok := result.Metadata["detector"].(string); ok {
		detectorName = v
	}
	detectorLatency := result.TotalLatencyMs
	if len(result.DetectorResults) > 0 {
		detectorLatency = result.DetectorResults[0].LatencyMs
	}
	s.Metrics.ObserveScan(string(result.Decision), result.TotalLatencyMs, result.ChunksScanned, detectorName, result.Score, detectorLatency)
	_ = s.Scans.RecordScanMetric(ctx, string(result.Decision), result.TotalLatencyMs)

	var scan *store.SecurityScan
	if opts.Persist {
		shouldPersist := true
		if s.Settings != nil && s.Settings.AuditStoreMode != "all" {
			// In blocked_only mode (default), only persist threat incidents (BLOCK, REVIEW, ERROR)
			shouldPersist = result.Decision != apis.DecisionAllow
		}

		if shouldPersist {
			var promptText *string
			if s.Settings != nil && s.Settings.PersistPromptContent {
				promptText = &text
			}
			stored, err := s.Scans.CreateFromResult(ctx, result, digest, len(text), promptText, opts.UserID, opts.WorkerID)
			if err != nil {
				return result, nil, err
			}
			scan = stored
			slog.Info("security_scan_persisted", "scan_id", scan.ID.String(), "request_id", result.RequestID, "decision", result.Decision)
		}
	}
	return result, scan, nil
}

type InspectOptions struct {
	RequestID string
	Source    *string
	UserID    *uuid.UUID
	Model     *string
	Metadata  map[string]interface{}
	Persist   bool
	WorkerID  *string
}

func (s *SecurityService) ListDetectors() []map[string]string {
	var implemented []map[string]string
	for _, name := range detectors.Names() {
		entry := map[string]string{"name": name, "status": "implemented"}
		if name == "rule-based" {
			entry["warning"] = "rule-based is development/testing only"
		} else if name == "prompt-shield" {
			entry["warning"] = "loads models from huggingface.co/collections/robbypambudi/prompt-shield"
		} else {
			entry["warning"] = ""
		}
		implemented = append(implemented, entry)
	}
		prepared := []map[string]string{
		{"name": "flan-t5", "status": "prepared", "warning": "use prompt-shield detector with HF sidecar"},
		{"name": "transformer", "status": "prepared", "warning": "not implemented"},
		{"name": "remote", "status": "prepared", "warning": "not implemented"},
		{"name": "ensemble", "status": "prepared", "warning": "not implemented"},
	}
	return append(implemented, prepared...)
}

func (s *SecurityService) Evaluate(ctx context.Context, samples []apis.EvaluateSample, threshold *float64, targetFPRs []float64) apis.EvaluationReport {
	var scores []float64
	var labels []bool
	for _, sample := range samples {
		result, _, _ := s.Inspect(ctx, sample.Text, InspectOptions{Source: strPtr("evaluate"), Persist: false})
		scores = append(scores, result.Score)
		labels = append(labels, sample.IsInjection)
	}
	usedThreshold := s.Pipeline.Policy.Threshold()
	if threshold != nil {
		usedThreshold = *threshold
	}
	if len(targetFPRs) == 0 {
		targetFPRs = []float64{0.0005, 0.001, 0.005, 0.01}
	}
	return evaluation.EvaluateScores(labels, scores, usedThreshold, targetFPRs)
}

func strPtr(s string) *string { return &s }
